File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.Configuration;
using DevExpress.Xpo;
using DevExpress.Xpo.Metadata;
using DevExpress.Xpo.Metadata.Helpers;
using DevExpress.Xpo.DB;

namespace ObjectProperties {
    /// <summary>
    /// Summary description for Form1.
    /// </summary>
    public class Form1 : System.Windows.Forms.Form {
        private System.Windows.Forms.ComboBox comboBox1;
        private System.Windows.Forms.ListBox listBox1;
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;

        public Form1() {
            InitializeComponent();

        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose(bool disposing) {
            if(disposing) {
                if(components != null) {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent() {
            this.comboBox1 = new System.Windows.Forms.ComboBox();
            this.listBox1 = new System.Windows.Forms.ListBox();
            this.SuspendLayout();
            // 
            // comboBox1
            // 
            this.comboBox1.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.comboBox1.Location = new System.Drawing.Point(24, 8);
            this.comboBox1.Name = "comboBox1";
            this.comboBox1.Size = new System.Drawing.Size(272, 24);
            this.comboBox1.TabIndex = 0;
            this.comboBox1.SelectedValueChanged += new System.EventHandler(this.comboBox1_SelectedValueChanged);
            // 
            // listBox1
            // 
            this.listBox1.ItemHeight = 16;
            this.listBox1.Location = new System.Drawing.Point(24, 64);
            this.listBox1.Name = "listBox1";
            this.listBox1.Size = new System.Drawing.Size(272, 260);
            this.listBox1.TabIndex = 1;
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(6, 15);
            this.ClientSize = new System.Drawing.Size(312, 341);
            this.Controls.AddRange(new System.Windows.Forms.Control[] {
																		  this.listBox1,
																		  this.comboBox1});
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            this.ResumeLayout(false);

        }
        #endregion

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main() {
            string conn = MSSqlConnectionProvider.GetConnectionString("(local)", "Northwind");
            XpoDefault.DataLayer = XpoDefault.GetDataLayer(conn, AutoCreateOption.SchemaAlreadyExists);
            XpoDefault.Session.Dictionary.CollectClassInfos(typeof(Customer).Assembly);
            Application.Run(new Form1());
        }

        private void Form1_Load(object sender, System.EventArgs e) {
            foreach(XPClassInfo info in XpoDefault.Session.Dictionary.Classes) {
                if(info.IsPersistent && info.IsVisibleInDesignTime)
                    comboBox1.Items.Add(info);
            }
        }

        string[] GetObjectProperties(XPClassInfo classInfo) {
            if(classInfo != null)
                return GetObjectProperties(classInfo, new ArrayList());
            return new string[] { };
        }

        string[] GetObjectProperties(XPClassInfo xpoInfo, ArrayList processed) {
            if(processed.Contains(xpoInfo)) return new string[] { };
            processed.Add(xpoInfo);
            ArrayList result = new ArrayList();
            foreach(XPMemberInfo m in xpoInfo.PersistentProperties)
                if(!(m is ServiceField) && m.IsPersistent) {
                    result.Add(m.Name);
                    if(m.ReferenceType != null) {
                        string[] childProps = GetObjectProperties(m.ReferenceType, processed);
                        foreach(string child in childProps)
                            result.Add(string.Format("{0}.{1}", m.Name, child));
                    }
                }

            foreach(XPMemberInfo m in xpoInfo.CollectionProperties) {
                string[] childProps = GetObjectProperties(m.CollectionElementType, processed);
                foreach(string child in childProps)
                    result.Add(string.Format("{0}.{1}", m.Name, child));
            }
            return result.ToArray(typeof(string)) as string[];
        }

        private void comboBox1_SelectedValueChanged(object sender, System.EventArgs e) {
            listBox1.Items.Clear();
            listBox1.Items.AddRange(GetObjectProperties((XPClassInfo)comboBox1.SelectedItem));
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Metadata
Imports DevExpress.Xpo.Metadata.Helpers
Imports DevExpress.Xpo.DB

Namespace ObjectProperties

    ''' <summary>
    ''' Summary description for Form1.
    ''' </summary>
    Public Class Form1
        Inherits Form

        Private comboBox1 As ComboBox

        Private listBox1 As ListBox

        ''' <summary>
        ''' Required designer variable.
        ''' </summary>
        Private components As System.ComponentModel.Container = Nothing

        Public Sub New()
            InitializeComponent()
        End Sub

        ''' <summary>
        ''' Clean up any resources being used.
        ''' </summary>
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If components IsNot Nothing Then
                    components.Dispose()
                End If
            End If

            MyBase.Dispose(disposing)
        End Sub

#Region "Windows Form Designer generated code"
        ''' <summary>
        ''' Required method for Designer support - do not modify
        ''' the contents of this method with the code editor.
        ''' </summary>
        Private Sub InitializeComponent()
            comboBox1 = New ComboBox()
            listBox1 = New ListBox()
            Me.SuspendLayout()
            ' 
            ' comboBox1
            ' 
            comboBox1.DropDownStyle = ComboBoxStyle.DropDownList
            comboBox1.Location = New System.Drawing.Point(24, 8)
            comboBox1.Name = "comboBox1"
            comboBox1.Size = New System.Drawing.Size(272, 24)
            comboBox1.TabIndex = 0
            AddHandler comboBox1.SelectedValueChanged, New EventHandler(AddressOf comboBox1_SelectedValueChanged)
            ' 
            ' listBox1
            ' 
            listBox1.ItemHeight = 16
            listBox1.Location = New System.Drawing.Point(24, 64)
            listBox1.Name = "listBox1"
            listBox1.Size = New System.Drawing.Size(272, 260)
            listBox1.TabIndex = 1
            ' 
            ' Form1
            ' 
            AutoScaleBaseSize = New System.Drawing.Size(6, 15)
            ClientSize = New System.Drawing.Size(312, 341)
            Me.Controls.AddRange(New Control() {listBox1, comboBox1})
            Name = "Form1"
            Text = "Form1"
            AddHandler Load, New EventHandler(AddressOf Form1_Load)
            Me.ResumeLayout(False)
        End Sub

#End Region
        ''' <summary>
        ''' The main entry point for the application.
        ''' </summary>
        <STAThread>
        Shared Sub Main()
            Dim conn As String = MSSqlConnectionProvider.GetConnectionString("(local)", "Northwind")
            XpoDefault.DataLayer = XpoDefault.GetDataLayer(conn, AutoCreateOption.SchemaAlreadyExists)
            XpoDefault.Session.Dictionary.CollectClassInfos(GetType(Customer).Assembly)
            Call Application.Run(New Form1())
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            For Each info As XPClassInfo In XpoDefault.Session.Dictionary.Classes
                If info.IsPersistent AndAlso info.IsVisibleInDesignTime Then comboBox1.Items.Add(info)
            Next
        End Sub

        Private Function GetObjectProperties(ByVal classInfo As XPClassInfo) As String()
            If classInfo IsNot Nothing Then Return GetObjectProperties(classInfo, New ArrayList())
            Return New String() {}
        End Function

        Private Function GetObjectProperties(ByVal xpoInfo As XPClassInfo, ByVal processed As ArrayList) As String()
            If processed.Contains(xpoInfo) Then Return New String() {}
            processed.Add(xpoInfo)
            Dim result As ArrayList = New ArrayList()
            For Each m As XPMemberInfo In xpoInfo.PersistentProperties
                If Not(TypeOf m Is ServiceField) AndAlso m.IsPersistent Then
                    result.Add(m.Name)
                    If m.ReferenceType IsNot Nothing Then
                        Dim childProps As String() = GetObjectProperties(m.ReferenceType, processed)
                        For Each child As String In childProps
                            result.Add(String.Format("{0}.{1}", m.Name, child))
                        Next
                    End If
                End If
            Next

            For Each m As XPMemberInfo In xpoInfo.CollectionProperties
                Dim childProps As String() = GetObjectProperties(m.CollectionElementType, processed)
                For Each child As String In childProps
                    result.Add(String.Format("{0}.{1}", m.Name, child))
                Next
            Next

            Return TryCast(result.ToArray(GetType(String)), String())
        End Function

        Private Sub comboBox1_SelectedValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            listBox1.Items.Clear()
            listBox1.Items.AddRange(GetObjectProperties(CType(comboBox1.SelectedItem, XPClassInfo)))
        End Sub
    End Class
End Namespace
